packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/MediaLibraryModule.kt

```diff
@@ -0,0 +1,467 @@
+
+  private var mImagesObserver: MediaStoreContentObserver? = null
+  private var mVideosObserver: MediaStoreContentObserver? = null
+  private var mAction: Action? = null
+
+  override fun getName() = "ExponentMediaLibrary"
+
+  override fun getConstants(): Map<String, Any> {
+    return mapOf(
+      "MediaType" to mapOf<String, Any>(
+        "audio" to MEDIA_TYPE_AUDIO,
+        "photo" to MEDIA_TYPE_PHOTO,
+        "video" to MEDIA_TYPE_VIDEO,
+        "unknown" to MEDIA_TYPE_UNKNOWN,
+        "all" to MEDIA_TYPE_ALL,
+      ),
+      "SortBy" to mapOf<String, Any>(
+        "default" to SORT_BY_DEFAULT,
+        "creationTime" to SORT_BY_CREATION_TIME,
+        "modificationTime" to SORT_BY_MODIFICATION_TIME,
+        "mediaType" to SORT_BY_MEDIA_TYPE,
+        "width" to SORT_BY_WIDTH,
+        "height" to SORT_BY_HEIGHT,
+        "duration" to SORT_BY_DURATION,
+      ),
+      "CHANGE_LISTENER_NAME" to LIBRARY_DID_CHANGE_EVENT,
+    )
+  }
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    moduleRegistryDelegate.onCreate(moduleRegistry)
+  }
+
+  @ExpoMethod
+  fun requestPermissionsAsync(writeOnly: Boolean, promise: Promise) {
+    Permissions.askForPermissionsWithPermissionsManager(
+      mPermissions,
+      promise,
+      *getManifestPermissions(writeOnly)
+    )
+  }
+
+  @ExpoMethod
+  fun getPermissionsAsync(writeOnly: Boolean, promise: Promise) {
+    Permissions.getPermissionsWithPermissionsManager(
+      mPermissions,
+      promise,
+      *getManifestPermissions(writeOnly)
+    )
+  }
+
+  @ExpoMethod
+  fun saveToLibraryAsync(localUri: String, promise: Promise) {
+    if (isMissingWritePermission) {
+      promise.reject(ERROR_NO_PERMISSIONS, ERROR_NO_WRITE_PERMISSION_MESSAGE)
+      return
+    }
+    CreateAsset(mContext, localUri, promise, false)
+      .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
+  }
+
+  @ExpoMethod
+  fun createAssetAsync(localUri: String, promise: Promise) {
+    if (isMissingPermissions) {
+      promise.reject(ERROR_NO_PERMISSIONS, ERROR_NO_PERMISSIONS_MESSAGE)
+      return
+    }
+    CreateAsset(mContext, localUri, promise)
+      .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
+  }
+
+  @ExpoMethod
+  fun addAssetsToAlbumAsync(
+    assetsId: List<String>,
+    albumId: String,
+    copyToAlbum: Boolean,
+    promise: Promise
+  ) {
+    if (isMissingPermissions) {
+      promise.reject(ERROR_NO_PERMISSIONS, ERROR_NO_PERMISSIONS_MESSAGE)
+      return
+    }
+    val action = Action action@{ permissionsWereGranted ->
+      if (!permissionsWereGranted) {
+        promise.reject(ERROR_NO_PERMISSIONS, ERROR_USER_DID_NOT_GRANT_WRITE_PERMISSIONS_MESSAGE)
+        return@action
+      }
+      AddAssetsToAlbum(mContext, assetsId.toTypedArray(), albumId, copyToAlbum, promise)
+        .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
+    }
+    runActionWithPermissions(if (copyToAlbum) emptyList() else assetsId, action, promise)
+  }
```

Comment:
I give under your consideration following way to refractor every method, that checks whether appropriate permissions were granted.
We could define additional private method `rejectUnlessPermissionsGranted`:

```kotlin
private inline fun rejectUnlessPermissionsGranted(promise: Promise, block: () -> Unit) {
  if (isMissingPermissions) {
    promise.reject(ERROR_NO_PERMISSIONS, ERROR_NO_PERMISSIONS_MESSAGE)
    return
  }
  block()
}
```

and write this method (and similar ones) like follows:

```kotlin
@ExpoMethod
fun addAssetsToAlbumAsync(
  assetsId: List<String>,
  albumId: String,
  copyToAlbum: Boolean,
  promise: Promise
) = rejectUnlessPermissionsGranted(promise) {
  <rest of the function body>
}
```

There would be need for another method: `rejectUnlessWritePermissionsGranted`, but IMHO it is nicer this way.

Or maybe we could create single method, with additional parameter describing what permissions should be verified -- this would simplify [possible changes (#10653)](https://github.com/expo/expo/issues/10653).